Offline cask search and listing all available casks

laughedelic opened the pull request:
Fixes #3080 

---- 

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
  _See below_
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Here are two changes for fixing #3080:
* a5640fd handles Github request failure, so that the whole search doesn't fail and returns empty "remote matches" instead
* 00803b1 makes `brew cask search` to output the list of all cask tokens without performing any search (analogously to `brew search`)

I tried to improve tests, but I have some problems with it and decided to submit a WIP pull-request and ask for help here:

1. I noticed that there was already a test for the expected behaviour described in #3080: ["lists all available Casks with no search term"](https://github.com/Homebrew/brew/blob/a5640fdfeeea7347e51c14abac481f4b210527ab/Library/Homebrew/test/cask/cli/search_spec.rb#L31-L35). But it fails **only** when I run tests with `--online` _and_ turned off network connection. I don't really understand why it passes without `--online` and no network. Are any network requests silently ignored without `--online` flag?
2. I noticed usage of `:needs_network` in other tests, but I don't know how to use it here: there should be tests both with and without network
3. Is there a way to specify that a test has to run successfully _without_ network? `:doesnt_need_network`? or a way to mock such situation?

reitermarkus commented on Library/Homebrew/cask/lib/hbc/cli/search.rb:
```diff
@@ -2,8 +2,12 @@ module Hbc
-        results = self.class.search(*args)
-        self.class.render_results(*results)
+        if args.empty?
+          puts Formatter.columns(CLI.nice_listing(Hbc.all_tokens))
```
This should only be called if there is no connection and should probably be handled in `::search`.

laughedelic replied:
The whole point of this change is that `brew cask search` (without arguments) _should not_ perform online search, same way is `brew search` does (see [here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/search.rb#L26-L27)). 
You can still do `brew cask search ""` which will do a normal search returning all casks.

reitermarkus commented on Library/Homebrew/cask/lib/hbc/cli/search.rb:
```diff
@@ -2,8 +2,12 @@ module Hbc
-        results = self.class.search(*args)
-        self.class.render_results(*results)
+        if args.empty?
+          puts Formatter.columns(CLI.nice_listing(Hbc.all_tokens))
```
This should only be called if there is no connection and should probably be handled in `::search`.

reitermarkus replied:
The last thing here would be to update the man page to mention this, other than that it looks good to me. (Make sure to run `brew man` afterwards.)

laughedelic commented on Library/Homebrew/cask/lib/hbc/cli/search.rb:
```diff
@@ -15,8 +19,18 @@ def self.extract_regexp(string)
+          filename: query,
+          extension: "rb"
+        )
+        rescue Exception => e
```
I changed it to `StandardError`. The error raised is [this](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/github.rb#L220-L226), can it be made more specific?

reitermarkus replied:
Ok, use `Github::Error` then.

laughedelic replied:
Done in 2cc6ef4

laughedelic commented on Library/Homebrew/cask/lib/hbc/cli/search.rb:
```diff
@@ -15,8 +19,18 @@ def self.extract_regexp(string)
       def self.search_remote(query)
-        matches = GitHub.search_code(user: "caskroom", path: "Casks",
-                                     filename: query, extension: "rb")
+        matches = begin GitHub.search_code(
```
15d2a9c like this?

reitermarkus commented:
Ok, now for testing his, you want `allow(GitHub).to receive(:search_code).and raise_error(GitHub::Error.new("curl failed! reason“))`, followed by the actual test.

laughedelic commented:
@reitermarkus hi! I added 2 new tests:
* for when online search failed: check the expected output and expected error message
* for empty matches output: it should be empty when stdout is not TTY

And modified the test for casks listing when `brew cask search` is called with no search term to ensure it's not affected by an online search failure.

I also unified formatting for some other tests in 68c837f, I hope you don't mind.

laughedelic commented on Library/Homebrew/cask/lib/hbc/cli/search.rb:
```diff
@@ -15,8 +19,18 @@ def self.extract_regexp(string)
+          )
+        rescue GitHub::Error => e
+          onoe e
+          $stderr.puts e.backtrace
```
One doubt here: probably outputting the backtrace is a bit too much? `onoe e` will output in this case 

```
Error: curl failed!
curl: (6) Could not resolve host: api.github.com
curl: (3) <url> malformed
```

And the backtrace is not helpful for the user:

```
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:226:in `raise_api_error'
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:184:in `open'
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:284:in `search'
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:239:in `search_code'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/search.rb:23:in `search_remote'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/search.rb:58:in `search'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/search.rb:8:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/abstract_command.rb:35:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:97:in `run_command'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:167:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:131:in `run'
/usr/local/Homebrew/Library/Homebrew/cmd/cask.rb:8:in `cask'
/usr/local/Homebrew/Library/Homebrew/brew.rb:95:in `<main>'
```

----

Probably the error message could be improved too (in `Library/Homebrew/utils/github.rb`). Like matching that error output from `curl`, checking the network connection and then doing `onoe 'Check your network connection'`. What do you think?

laughedelic replied:
I removed the stacktrace in 2a3f83b8d4ceeca370a60bbcb494990a2bd20801 and changed error to warning (as it doesn't fail the whole search). Now it looks like this:

```
$ brew cask search "foo-bar"
Warning: Online search failed: curl failed!
curl: (6) Could not resolve host: api.github.com
curl: (3) <url> malformed

No Cask found for "foo-bar".
```

Online/Github/Remote search?

laughedelic commented:
@reitermarkus could you review this again?

reitermarkus commented on Library/Homebrew/cask/lib/hbc/cli/search.rb:
```diff
@@ -15,8 +19,17 @@ def self.extract_regexp(string)
+            extension: "rb",
+          )
+        rescue GitHub::Error => error
+          opoo "Online search failed: #{error}\n"
```
I'd rephrase this to “Error searching on GitHub:”.

laughedelic replied:
Done in 68dd0ac

reitermarkus commented on Library/Homebrew/test/cask/cli/search_spec.rb:
```diff
@@ -22,16 +22,38 @@
+  it "doesn't output anything to non-TTY stdout when there are no matches" do
+    expect {
+      Hbc::CLI::Search.run("foo-bar-baz")
+    }.to not_to_output.to_stdout
```
`.not_to output.to_stdout`

laughedelic replied:
I tried to change it to `.not_to output.to_stdout.or output.to_stderr`, but then got the following error:

> `expect(...).not_to matcher.or matcher` is not supported, since it creates a bit of an ambiguity. Instead, define negated versions of whatever matchers you wish to negate with `RSpec::Matchers.define_negated_matcher` and use `expect(...).to matcher.or matcher`.

There are two matchers there: no output to stdout and no output to stderr. How should I rewrite it?

reitermarkus replied:
Ok, then make it `.to not_output.to_stderr`, and add another negated matcher like `not_to_output`.

laughedelic replied:
Sorry, I'm not sure what you mean. It seems that it is already as you say, two negated conditions: `¬A & ¬B`:

```ruby
.to not_to_output.to_stdout
.and not_to_output.to_stderr
```

I changed the formatting in 8e5ad9a though.

reitermarkus commented on Library/Homebrew/test/cask/cli/search_spec.rb:
```diff
@@ -22,16 +22,38 @@
     expect {
       Hbc::CLI::Search.run
     }.to output(/local-caffeine/).to_stdout.as_tty
+    .and not_to_output.to_stderr
```
Make the expect a single line and put the `.to` and `.and` line below it.

laughedelic replied:
Done in 8e5ad9a. Should I change the rest of the tests to this formatting too?

laughedelic commented:
@reitermarkus done in 50db7db 👌

reitermarkus commented:
Great work here, @laughedelic!

laughedelic commented:
Thanks!
